[<img src="img/DSunimib.png", width = 910, height = 520>](http://datascience.disco.unimib.it/it/)

<br>

#  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<font color=C21212>Foundations of Computer Science (FoCS)  -  Project</font>

***

> - __ID number__: &nbsp;&nbsp;800069
- __Surname__: &nbsp;&nbsp;_Borroni_&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;__Name__: &nbsp;&nbsp;_Alessandro_
- __Academic year__: &nbsp;&nbsp;2018/2019
- __CdLM__: &nbsp;&nbsp;Data Science
- __Faculty__: &nbsp;&nbsp;Department of Informatics, Systems and Communication (DISCO)
- __Class__: &nbsp;&nbsp;LM-91 - Classe delle Lauree Magistrali in Tecniche e metodi per la Società dell'Informazione
- __E-mail__: &nbsp;&nbsp;a.borroni2@campus.unimib.it

***

## <font color=BF4100> Table of contents </font>
> &nbsp; &#9679; &nbsp;&nbsp; [Introduction](#paragraph1) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9675; &nbsp;&nbsp; [Links to the instructions and datasets used for the project](#subparagraph1) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9675; &nbsp;&nbsp; [Import libraries](#subparagraph2) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9675; &nbsp;&nbsp; [Import and explore datasets](#subparagraph3) <br>
&nbsp; &#9679; &nbsp;&nbsp; [Solution to the exercises](#paragraph2) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1) &nbsp;&nbsp; [Convert the app sizes to a number](#ex1) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2) &nbsp;&nbsp; [Convert the number of installs to a number](#ex2) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3) &nbsp;&nbsp; [Transform “Varies with device” into a missing value](#ex3) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 4) &nbsp;&nbsp; [Convert Current Ver and Android Ver into a dotted number](#ex4) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 5) &nbsp;&nbsp; [Remove the duplicates](#ex5) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 6) &nbsp;&nbsp; [For each category, compute the number of apps](#ex6) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 7) &nbsp;&nbsp; [For each category, compute the average rating](#ex7) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 8) &nbsp;&nbsp; [Create two dataframes: one for the genres and one bridging apps and genres](#ex8) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 9) &nbsp;&nbsp; [For each genre, create a new column of the original dataframe](#ex9) <br>
    &nbsp;&nbsp;&nbsp; 10) &nbsp;&nbsp; [For each genre, compute the average rating](#ex10) <br>
    &nbsp;&nbsp;&nbsp; 11) &nbsp;&nbsp; [For each app, compute the approximate income, obtain as a product of number of installs and price](#ex11) <br>
    &nbsp;&nbsp;&nbsp; 12) &nbsp;&nbsp; [For each app, compute its minimum and maximum ` Sentiment_polarity `](#ex12) <br>
   &nbsp; &#9679; &nbsp;&nbsp; [Conclusion](#paragraph3) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9675; &nbsp;&nbsp; [First "_time_" ending](#subparagraph4) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9675; &nbsp;&nbsp; [Some Data Visualizations](#subparagraph5) <br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9675; &nbsp;&nbsp; [Second "_time_" ending](#subparagraph6)


<a name="paragraph1"></a>

## <font color=BF4100> Introduction </font>

In questa fase introduttiva svolgo qualche passaggio basilare prima di approcciarmi al progetto in sè e alle specifiche richieste oggetto di valutazione, come l'importazione delle librerie necessarie e l'esplorazione dei dataset. <br>
In più, ho deciso di allegare qualche link utile al fine di avere un'idea più precisa del progetto in sè e degli stessi dataset.

<a name="subparagraph1"></a>

  ### <p>&#9656;  <font color=E26200> Links to instructions and datasets used for the project</font> </p>
 

Per le istruzioni in merito al progetto, cliccare [qui](https://gdv.github.io/foundationsCS-2018/project "Clicca qui per vedere le istruzioni.").

[<img src="img/GPS.png", width = 300, height = 200>](https://play.google.com/store?hl=it)

Per scaricare i dataset, cliccare sui seguenti link:
<br> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; i) &nbsp;&nbsp;&nbsp; [Google Play Store](https://raw.githubusercontent.com/gdv/foundationsCS-2018/master/ex-data/project/googleplaystore.csv)
<br> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ii) &nbsp;&nbsp; [Google Play Store: User reviews](https://raw.githubusercontent.com/gdv/foundationsCS-2018/master/ex-data/project/googleplaystore_user_reviews.csv)

<a name="subparagraph2"></a>

  ### <p>&#9656;  <font color=E26200> Import *libraries*</font> </p>

In [1]:
import pandas as pd
import numpy as np
import re
import time

Uso "*time*" con l'intento di avere un resoconto sul tempo di esecuzione del codice.

In [2]:
begin = time.time()

Il risultato, ovviamente, verrà mostrato in fondo al Notebook, quando la funzione verrà chiamata nuovamente nella sua fase conclusiva.

<a name="subparagraph3"></a>

   ### <p>&#9656;  <font color=E26200> Import and explore *datasets*</font> </p>
   

In [3]:
play_store = pd.read_csv('C:/Users/MioPC/Documents/googleplaystore.csv')
user_reviews = pd.read_csv('C:/Users/MioPC/Documents/googleplaystore_user_reviews.csv')

Decido di controllare il dataset denominato: "*googleplaystore.csv*", limitando l'output alle prime 5 righe.

In [4]:
play_store.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

Decido di controllare il dataset denominato: "*googleplaystore_user_reviews.csv*", sempre limitando l'output a 5 righe.

In [5]:
user_reviews.head()

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000

Eseguo un controllo sul *dtype* degli attributi.

In [6]:
print(play_store.dtypes)

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object


In [7]:
print(user_reviews.dtypes)

App                        object
Translated_Review          object
Sentiment                  object
Sentiment_Polarity        float64
Sentiment_Subjectivity    float64
dtype: object


Dopo un'attenta analisi, mi rendo conto che il record *_10472_* presenta dei valori inaccurati.<br> Più precisamente, la mancanza di un valore per il campo "*Category*" causa uno spostamento di tutti gli altri valori a sinistra di una posizione.<br> Qui sotto si può verificare quanto appena affermato:

In [8]:
play_store.loc[[10472]]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

La cosa migliore da fare, quindi, è eliminare tale record.

In [9]:
play_store = play_store.drop(10472)

***

<a name="paragraph2"></a>

## <font color=BF4100> Solution to the exercises </font>

## <font color=000080> Exercise 1)</font> </p>
<a name="ex1"></a>

> ### <font color=0060A0> Convert the app sizes to a number</font> </p>

Prima di proseguire, controllo il formato dei dati:

In [10]:
play_store.Size.unique()[0:8]

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M'],
      dtype=object)

Al fine di evitare problemi con il record che presenta il valore "*Varies with device*" nel campo relativo alla Size, scrivo un pezzo di codice che mi permette di convertirlo in un numero che sicuramente non verrà preso in considerazione dalla funzione creata.<br> Scelgo di usare il valore "999999999".

In [11]:
play_store['Size'] = [re.sub('Varies with device', '999999999', size) for size in play_store['Size']]
play_store['Size'] = [re.sub(',', '.', size) for size in play_store['Size']]

Nella seconda linea di codice ho sostituito un punto alla virgola.

In [12]:
size_to_number = re.compile('(?P<amount>\d*\.*\d*)(?P<unit>\w*\+*)')

def unit_to_mult(unit):
    if unit == 'G':
        return 1000000000
    if unit == 'M':
        return 1000000
    if unit == 'k':
        return 1000
    return 1

Nella funzione sopraindicata si usano rispettivamente:<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; i) *k* &nbsp;per indicare le migliaia (*kilo*);<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ii) *M* &nbsp;per indicare i milioni (*Mega*);<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; iii) *G* &nbsp;per indicare i miliardi (*Giga*).<br>

In [13]:
def to_numeric(elem):
    m = size_to_number.search(elem)
    unit = m.group('unit')
    mult = unit_to_mult(unit)
    amount = float(m.group('amount'))
    return int(amount * mult)

La funzione appena scritta ci permette di convertire le stringhe in numeri.

In [14]:
play_store['Standard size'] = play_store['Size'].apply(to_numeric)

Dopo aver applicato le trasformazioni necessarie su "*Size*", posso finalmente riportare "*Varies with device*" nella sua forma originaria. Questo al fine di non alterare il significato iniziale dell'attributo.

In [15]:
play_store['Size'] = [re.sub('999999999', 'Varies with device', size) for size in play_store['Size']]
play_store['Standard size'] = play_store['Standard size'].apply(str)
play_store['Standard size'] = [re.sub('999999999', 'Varies with device', size) for size in play_store['Standard size']]

play_store.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver Standard size  
0  4.0.3 and up      19000000  
1  4.0.3 and up      14000000  
2  4.0.3 and up       8700000  
3    4.2 and up      25000000  
4    4.4 and up       2800000

Sposto la colonna '*Size*' in fondo, mediante l'ausilio di una List Comprehension, così da permettere un confronto immediato con la colonna '*Standard size*' appena creata.

In [16]:
play_store = play_store[[i for i in play_store if i not in ['Size', 'Standard size']] + ['Size', 'Standard size']]

play_store.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews     Installs  Type Price Content Rating                     Genres  \
0     159      10,000+  Free     0       Everyone               Art & Design   
1     967     500,000+  Free     0       Everyone  Art & Design;Pretend Play   
2   87510   5,000,000+  Free     0       Everyone               Art & Design   
3  215644  50,000,000+  Free     0           Teen               Art & Design   
4     967     100,000+  Free     0       Everyone    Art & Design;Creativity   

       Last Updated         Current Ver   Android Ver  Size Standard size  
0   January 7, 2018               1.0.0  4.0.3 and up   19M      19000000  
1  January 15, 2018               2.0.0  4.0.3 and up   14M      14000000  
2    August 1, 2018               1.2.4  4.0.3 and up  8.7M       8700000  
3      June 8, 2018  Varies with device    4.2 and up   25M      25000000  
4     June 20, 2018                 1.1    4.4 and up  2.8M       2800000

## <font color=000080> Exercise 2)</font> </p>
<a name="ex2"></a>

> ### <font color=0060A0> Convert the number of installs to a number</font> </p>

Ancora una volta, prima di proseguire controllo il formato dei dati:

In [17]:
play_store.Installs.unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+', '0'], dtype=object)

Identifico eventuali valori anomali in maniera più precisa attraverso una Group By.

In [18]:
play_store.groupby('Installs')['Installs'].size()

Installs
0                    1
0+                  14
1+                  67
1,000+             907
1,000,000+        1579
1,000,000,000+      58
10+                386
10,000+           1054
10,000,000+       1252
100+               719
100,000+          1169
100,000,000+       409
5+                  82
5,000+             477
5,000,000+         752
50+                205
50,000+            479
50,000,000+        289
500+               330
500,000+           539
500,000,000+        72
Name: Installs, dtype: int64

L'unica riga diversa dalle altre è quella che presenta uno " *0* " senza il simbolo " *+* " alla sua destra.

In [19]:
display(play_store[play_store.Installs == '0'])

App Category  Rating Reviews Installs Type Price  \
9148  Command & Conquer: Rivals   FAMILY     NaN       0        0  NaN     0   

     Content Rating    Genres   Last Updated         Current Ver  \
9148   Everyone 10+  Strategy  June 28, 2018  Varies with device   

             Android Ver                Size       Standard size  
9148  Varies with device  Varies with device  Varies with device

Tuttavia questo non costituisce un problema, dal momento che il summenzionato simbolo sarà ignorato nello svolgimento dell'esercizio.

In [20]:
play_store['N° of Installs'] = [re.sub('\\D', '', number) for number in play_store['Installs']]
play_store['N° of Installs'] = play_store['N° of Installs'].apply(int)

play_store['N° of Installs'].dtypes

dtype('int64')

In ordine:<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; i) Rimuovo tutti i caratteri dagli attributi;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ii) Converto i valori in interi;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; iii) Controllo il tipo per avere una conferma dell'avvenuta modificazione.

In [21]:
play_store = play_store[[i for i in play_store if i not in ['Installs', 'N° of Installs']] + ['Installs', 'N° of Installs']]

play_store.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Type Price Content Rating                     Genres  \
0     159  Free     0       Everyone               Art & Design   
1     967  Free     0       Everyone  Art & Design;Pretend Play   
2   87510  Free     0       Everyone               Art & Design   
3  215644  Free     0           Teen               Art & Design   
4     967  Free     0       Everyone    Art & Design;Creativity   

       Last Updated         Current Ver   Android Ver  Size Standard size  \
0   January 7, 2018               1.0.0  4.0.3 and up   19M      19000000   
1  January 15, 2018               2.0.0  4.0.3 and up   14M      14000000   
2    August 1, 2018               1.2.4  4.0.3 and up  8.7M       8700000   
3      June 8, 2018  Varies with device    4.2 and up   25M      25000000   
4     June 20, 2018                 1.1    4.4 and up  2.8M       2800000   

      Installs  N° of Installs  
0      10,000+           10000  
1     500,000+          500000  
2   5,000,000+         5000000  
3  50,000,000+        50000000  
4     100,000+          100000

Ho eseguito una List Comprehension al fine di affiancare le colonne di interesse in fondo alla tabella.

## <font color=000080> Exercise 3)</font> </p>
<a name="ex3"></a>

> ### <font color=0060A0> Transform “Varies with device” into a missing value</font> </p>

Trasformo "*Varies with device*" in un missing value, rimpiazzandolo con un NaN.

In [22]:
play_store_nan = play_store.replace('Varies with device', np.nan)

Dopo aver utilizzato la trasformazione in NaN di "*Varies with device*", posso finalmente convertire l'attributo "*Standard size*" in un float.

In [23]:
play_store_nan['Standard size'] = play_store_nan['Standard size'].apply(float)

play_store_nan['Standard size'].dtypes

dtype('float64')

## <font color=000080> Exercise 4)</font> </p>
<a name="ex4"></a>

> ### <font color=0060A0> Convert Current Ver and Android Ver into a dotted number (e.g. 4.0.3 or 4.2)</font> </p>

Dò una rapida occhiata a un estratto del dataset, focalizzandomi sulle colonne specificate nella consegna dell'esercizio:

In [24]:
play_store_nan[['App', 'Current Ver', 'Android Ver']].head(10)

App Current Ver   Android Ver
0     Photo Editor & Candy Camera & Grid & ScrapBook       1.0.0  4.0.3 and up
1                                Coloring book moana       2.0.0  4.0.3 and up
2  U Launcher Lite – FREE Live Cool Themes, Hide ...       1.2.4  4.0.3 and up
3                              Sketch - Draw & Paint         NaN    4.2 and up
4              Pixel Draw - Number Art Coloring Book         1.1    4.4 and up
5                         Paper flowers instructions         1.0    2.3 and up
6            Smoke Effect Photo Maker - Smoke Editor         1.1  4.0.3 and up
7                                   Infinite Painter    6.1.61.1    4.2 and up
8                               Garden Coloring Book       2.9.2    3.0 and up
9                      Kids Paint Free - Drawing Fun         2.8  4.0.3 and up

Decido di privare le colonne "*Current Ver*" e "*Android Ver*" di tutti i caratteri non numerici.

In [25]:
play_store_nan['Current Ver Refined'] = play_store_nan['Current Ver']
play_store_nan['Android Ver Refined'] = play_store_nan['Android Ver']

# Rimuovo caratteri non numerici da "Current Ver":
play_store_nan['Current Ver Refined'].replace('[a-z]+|[A-z]+', '', regex = True, inplace = True)

# Rimuovo caratteri non numerici da "Android Ver":
play_store_nan['Android Ver Refined'].replace('[a-z]+|[A-z]+', '', regex = True, inplace = True)

Rimuovo dalle due versioni anche eventuali punti che non sono seguiti da una cifra numerica:

In [26]:
play_store_nan['Current Ver Refined'].replace(r'(\.+$)', '', regex = True, inplace = True)
play_store_nan['Android Ver Refined'].replace(r'(\.+$)', '', regex = True, inplace = True)

play_store_nan[['App', 'Current Ver', 'Current Ver Refined', 'Android Ver', 'Android Ver Refined']].head()

App Current Ver  \
0     Photo Editor & Candy Camera & Grid & ScrapBook       1.0.0   
1                                Coloring book moana       2.0.0   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...       1.2.4   
3                              Sketch - Draw & Paint         NaN   
4              Pixel Draw - Number Art Coloring Book         1.1   

  Current Ver Refined   Android Ver Android Ver Refined  
0               1.0.0  4.0.3 and up             4.0.3    
1               2.0.0  4.0.3 and up             4.0.3    
2               1.2.4  4.0.3 and up             4.0.3    
3                 NaN    4.2 and up               4.2    
4                 1.1    4.4 and up               4.4

Volendo mostrare il dataset nella sua interezza, mi avvalgo, come nei punti precedenti, di una List Comprehension al fine di avere le colonne interessate in fondo alla tabella di output:

In [27]:
play_store_nan = play_store_nan[[i for i in play_store_nan if i not in ['Current Ver', 'Current Ver Refined', 'Android Ver', 'Android Ver Refined']] + ['Current Ver', 'Current Ver Refined', 'Android Ver', 'Android Ver Refined']]

play_store_nan.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Type Price Content Rating                     Genres  \
0     159  Free     0       Everyone               Art & Design   
1     967  Free     0       Everyone  Art & Design;Pretend Play   
2   87510  Free     0       Everyone               Art & Design   
3  215644  Free     0           Teen               Art & Design   
4     967  Free     0       Everyone    Art & Design;Creativity   

       Last Updated  Size  Standard size     Installs  N° of Installs  \
0   January 7, 2018   19M     19000000.0      10,000+           10000   
1  January 15, 2018   14M     14000000.0     500,000+          500000   
2    August 1, 2018  8.7M      8700000.0   5,000,000+         5000000   
3      June 8, 2018   25M     25000000.0  50,000,000+        50000000   
4     June 20, 2018  2.8M      2800000.0     100,000+          100000   

  Current Ver Current Ver Refined   Android Ver Android Ver Refined  
0       1.0.0               1.0.0  4.0.3 and up             4.0.3    
1       2.0.0               2.0.0  4.0.3 and up             4.0.3    
2       1.2.4               1.2.4  4.0.3 and up             4.0.3    
3         NaN                 NaN    4.2 and up               4.2    
4         1.1                 1.1    4.4 and up               4.4

## <font color=000080> Exercise 5)</font> </p>
<a name="ex5"></a>

> ### <font color=0060A0> Remove the duplicates</font> </p>

L'esistenza dei duplicati è probabilmente dovuta alla fase di raccolta dei dati, avvenuta in tempi differenti.<br>
Per questo punto ho quindi deciso di ordinare il dataset in base agli attributi "*App*" e "*Reviews*", optando per un ordinamento decrescente degli stessi.<br>
Così facendo ho controllo su quali applicazioni abbiano ricevuto un numero minore di recensioni, potendo quindi procedere all'eliminazione dei duplicati e al mantenimento del primo record, più recente nonché più interessante (perché contenente maggiori informazioni).

In [28]:
print('Il numero di righe prima di aver effettuato ogni sorta di modifica è pari a', len(play_store_nan))

Il numero di righe prima di aver effettuato ogni sorta di modifica è pari a 10840


Volendo avrei potuto usare la funzione "*shape*" per avere un'informazione anche sul numero di colonne.

In [29]:
print('Le righe e le colonne totali sono pari a:', play_store_nan.shape)

Le righe e le colonne totali sono pari a: (10840, 17)


Cerco di capire se vi siano record completamente identici, usando "*drop_duplicates*".

In [30]:
play_store_nan.drop_duplicates(inplace = True)

print('Il numero di righe dopo aver eliminato i duplicati è pari a', len(play_store_nan))

Il numero di righe dopo aver eliminato i duplicati è pari a 10357


Vista la riduzione da 10840 righe a 10357, si può evincere come esistessero circa 483 righe coincidenti.

In [31]:
play_store_nan['Reviews'] = pd.to_numeric(play_store_nan['Reviews'])
play_store_nan[play_store_nan['App'] == 'Netflix'][['App', 'Reviews']]

App  Reviews
855   Netflix  5456208
931   Netflix  5456599
958   Netflix  5456708
3889  Netflix  5453997

Come applicazione di riferimento decido di utilizzare Netflix.<br>
Il passo successivo è quello di ordinare i valori dal maggiore al minore, mantenendo poi solo quello con il numero di recensioni più alto.

In [32]:
play_store_nan = play_store_nan.sort_values('Reviews', ascending = False)
play_store_nan[play_store_nan['App'] == 'Netflix'][['App', 'Reviews']]

App  Reviews
958   Netflix  5456708
931   Netflix  5456599
855   Netflix  5456208
3889  Netflix  5453997

In [33]:
play_store_nan.drop_duplicates(subset = 'App', keep = 'first', inplace = True)
play_store_nan[play_store_nan['App'] == 'Netflix'][['App', 'Reviews']]

App  Reviews
958  Netflix  5456708

In [34]:
print('La totalità di righe e colonne dopo aver svolto il punto del progetto risulta pari a:', play_store_nan.shape)

La totalità di righe e colonne dopo aver svolto il punto del progetto risulta pari a: (9659, 17)


## <font color=000080> Exercise 6)</font> </p>
<a name="ex6"></a>

> ### <font color=0060A0> For each category, compute the number of apps</font> </p>

Avendo già rimosso i duplicati al punto 5), mi basta fare una Group By e usare la funzione "*size*"; così facendo raggrupperò per l'attributo "*Category*" e conterò i valori uguali.

In [35]:
play_store_nan.groupby('Category').size().sort_values(ascending = True)

Category
BEAUTY                   53
COMICS                   56
PARENTING                60
ART_AND_DESIGN           61
EVENTS                   64
HOUSE_AND_HOME           73
WEATHER                  79
LIBRARIES_AND_DEMO       84
AUTO_AND_VEHICLES        85
ENTERTAINMENT            86
EDUCATION               105
FOOD_AND_DRINK          112
MAPS_AND_NAVIGATION     131
VIDEO_PLAYERS           164
DATING                  170
SHOPPING                202
TRAVEL_AND_LOCAL        219
BOOKS_AND_REFERENCE     222
SOCIAL                  239
NEWS_AND_MAGAZINES      254
PHOTOGRAPHY             281
HEALTH_AND_FITNESS      288
COMMUNICATION           315
SPORTS                  325
FINANCE                 345
LIFESTYLE               369
PRODUCTIVITY            374
PERSONALIZATION         376
MEDICAL                 395
BUSINESS                420
TOOLS                   829
GAME                    945
FAMILY                 1878
dtype: int64

## <font color=000080> Exercise 7)</font> </p>
<a name="ex7"></a>

> ### <font color=0060A0> For each category, compute the average rating</font> </p>

Per questo punto basta usare una semplice Group By accompagnata da "_.mean()_" per avere la media delle varie categorie (dal valore più alto al valore più basso).

In [36]:
play_store_nan.groupby('Category')[['Rating']].mean().sort_values(by = 'Rating', ascending = False).head(10)

Rating
Category                     
EVENTS               4.435556
ART_AND_DESIGN       4.359322
EDUCATION            4.349038
BOOKS_AND_REFERENCE  4.344970
PERSONALIZATION      4.332215
PARENTING            4.300000
BEAUTY               4.278571
SOCIAL               4.247291
GAME                 4.244432
WEATHER              4.243056

In alternativa, usando DataFrame di pandas, si possono ottenere gli stessi risultati ma con la colonna "_Category_" in ordine alfabetico.

In [37]:
pd.DataFrame(play_store_nan.groupby('Category')['Rating'].mean()).head(10)

Rating
Category                     
ART_AND_DESIGN       4.359322
AUTO_AND_VEHICLES    4.190411
BEAUTY               4.278571
BOOKS_AND_REFERENCE  4.344970
BUSINESS             4.098479
COMICS               4.181481
COMMUNICATION        4.121484
DATING               3.980451
EDUCATION            4.349038
ENTERTAINMENT        4.129070

## <font color=000080> Exercise 8)</font> </p>
<a name="ex8"></a>

> ### <font color=0060A0> Create two dataframes: one for the genres and one bridging apps and genres<br>So that, for instance, the app Pixel Draw - Number Art Coloring Book appears twice in the bridging table, once for Art & Design, once for Creativity</font> </p>

Cerco di capire come è strutturata la colonna "_Genres_", focalizzandomi sul numero di generi per ogni applicazione.

In [38]:
play_store_nan.groupby('Genres').size() 

Genres
Action                                   299
Action;Action & Adventure                 12
Adventure                                 73
Adventure;Action & Adventure               5
Adventure;Brain Games                      1
Adventure;Education                        1
Arcade                                   184
Arcade;Action & Adventure                 14
Arcade;Pretend Play                        1
Art & Design                              57
Art & Design;Action & Adventure            1
Art & Design;Creativity                    6
Art & Design;Pretend Play                  1
Auto & Vehicles                           85
Beauty                                    53
Board                                     42
Board;Action & Adventure                   3
Board;Brain Games                         14
Board;Pretend Play                         1
Books & Reference                        222
Books & Reference;Creativity               1
Books & Reference;Education                2
Bus

Noto che le applicazioni presentano al massimo due generi, ma con le seguenti differenze:<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; i) Alcune applicazioni presentano due generi identici

In [39]:
play_store_nan['Genres'].loc[2035]

'Education;Education'

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ii) Altre applicazioni presentano due generi differenti<br>

In [40]:
play_store_nan['Genres'].loc[3712]

'Video Players & Editors;Creativity'

Ma questo non rappresenta un vero e proprio problema, dal momento che in caso di più generi, questi verranno tenuti distinti nel nuovo dataframe, usando come riferimento la presenza di un punto e virgola (" ; ") tra due generi. <br>
Quindi non mi resta che creare i due dataframe richiesti, partendo dal primo, quello dei generi.

In [41]:
play_store_nan['Genre_Splitter'] = [gnr.split(';') for gnr in play_store_nan['Genres']]
genres = play_store_nan.Genre_Splitter.apply(pd.Series)

genres.tail(6)

0    1
10344             Tools  NaN
8872   Health & Fitness  NaN
8871       Art & Design  NaN
10347      Food & Drink  NaN
10348      House & Home  NaN
6329             Casino  NaN

In [42]:
genres = genres.rename(columns = {0: 'First/Main Genre', 1: 'Second Genre'})
genres.head(5)

First/Main Genre Second Genre
2544           Social          NaN
336     Communication          NaN
2604           Social          NaN
382     Communication          NaN
1879         Strategy          NaN

In ordine ho:<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; i) Eseguito lo splitting dei due generi in un vettore (array);<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ii) Creato il relativo dataframe;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; iii) Sistemato i nomi delle colonne.

Controllo ora una delle righe messe in evidenza precedentemente, per accertarmi dell'avvenuto splitting.

In [43]:
genres.loc[[3712]]

First/Main Genre Second Genre
3712  Video Players & Editors   Creativity

Prima di procedere alla creazione della tabella di collegamento, creo un dataframe che contiene i nomi delle applicazioni. <br>
Questa precisazione serve a far sì che non mi trovi a lavorare con una serie.

In [44]:
applications = play_store_nan['App'].to_frame()
type(applications)

pandas.core.frame.DataFrame

Nella creazione della tabella-ponte, uso "_.dropna()_" per eliminare i valori mancanti, e quindi per evitare che le applicazioni che non hanno un secondo genere si ripetano più volte.

In [45]:
bridging_table = pd.merge(genres, applications, left_index = True, right_index = True)\
                 .melt(id_vars = ['App'], value_name = 'Main Genre')\
                 .drop('variable', axis = 1)\
                 .dropna()
bridging_table.head(10)

App               Main Genre
0                                           Facebook                   Social
1                                 WhatsApp Messenger            Communication
2                                          Instagram                   Social
3           Messenger – Text and Video Chat for Free            Communication
4                                     Clash of Clans                 Strategy
5            Clean Master- Space Cleaner & Antivirus                    Tools
6                                     Subway Surfers                   Arcade
7                                            YouTube  Video Players & Editors
8  Security Master - Antivirus, VPN, AppLock, Boo...                    Tools
9                                       Clash Royale                 Strategy

Volendo fare due controlli:

In [46]:
bridging_table[bridging_table['App'] == 'Pixel Draw - Number Art Coloring Book']

App    Main Genre
4831   Pixel Draw - Number Art Coloring Book  Art & Design
14490  Pixel Draw - Number Art Coloring Book    Creativity

In [47]:
bridging_table[bridging_table['App'] == 'Video Editor']

App               Main Genre
1156   Video Editor  Video Players & Editors
10815  Video Editor               Creativity

Decido di esportare i dati in formato CSV e XLSX, per eventuali manipolazioni future.

In [48]:
bridging_table.to_csv("C:/Users/MioPC/Documents/GPS_bridging_table.csv")

In [49]:
bridging_table.to_excel("C:/Users/MioPC/Documents/GPS_bridging_table.xlsx")

## <font color=000080> Exercise 9)</font> </p>
<a name="ex9"></a>

> ### <font color=0060A0> For each genre, create a new column of the original dataframe <br> The new columns must have boolean values (True if the app has a given genre)</font> </p>

Per ottenere quanto richiesto dalla consegna ho fatto ricorso a una funzione, che ho chiamato "*generate_dummies*". <br>
L'idea di fondo è quella di creare delle colonne riferite ai diversi generi presenti nel dataset e di avere nelle celle relative a queste colonne un valore booleano, quindi o *True* o *False*, in base all'applicazione considerata. <br>
Bisogna precisare che in riferimento ai generi viene sfruttata la distinzione tra *First* e *Second Genre* elaborata durante l'**Exercise 8)**.

In [50]:
def generate_dummies(df, colname):
    boolean_clmn = pd.get_dummies(df[colname].apply(pd.Series).stack()).sum(level = 0)
    boolean_clmn.drop(boolean_clmn.columns[0], axis = 1, inplace = True)
    boole = pd.concat([df, boolean_clmn.astype(bool)], axis = 1)
    boole.drop(colname, axis = 1, inplace = True)
    return boole

Volendo fare un esempio per confermare l'avvenuta modifica:

In [51]:
is_that_genre = generate_dummies(play_store_nan, 'Genre_Splitter')
is_that_genre[is_that_genre['App'] == 'Video Editor'][['App', 'Genres', 'Creativity', 'Education', 'Art & Design', 'Video Players & Editors', 'Pretend Play', 'Finance', 'Strategy', 'Trivia']]

App                              Genres  Creativity  Education  \
3712  Video Editor  Video Players & Editors;Creativity        True      False   

      Art & Design  Video Players & Editors  Pretend Play  Finance  Strategy  \
3712         False                     True         False    False     False   

      Trivia  
3712   False

## <font color=000080> Exercise 10)</font> </p>
<a name="ex10"></a>

> ### <font color=0060A0> For each genre, compute the average rating <br>What is the genre with highest average?</font> </p>

La prima cosa che possiamo fare è controllare l'esistenza di valori mancanti nella colonna "_Rating_". <br>
Non comprometterà lo svolgimento di questo punto del progetto, ma è comunque meglio essere consapevoli dei valori con cui avremo a che fare.

In [52]:
play_store_nan['Rating'].isna().any()

True

Procedendo allo svolgimento del punto in esame, basterà unire il dataset originario (con le modifiche strutturali fino a qui applicate, ovviamente) con la tabella ponte ("*bridging_table*") creata durante l'**Exercise 9)**. <br>
In questo modo otteniamo una tabella contenente i rating delle varie applicazioni, da cui si può ottenere facilmente un rating medio per ogni genere. <br>
L'utilizzo della "*bridging_table*" viene giustificato dall'intento di preservare la distinzione dei generi tra _First_ e _Second_ rispetto alla codifica originaria data da "_Genres_".

In [53]:
rating_table = bridging_table.merge(play_store_nan, on = 'App')
rating_table[['App', 'Genres', 'Main Genre']]

App  \
0                                               Facebook   
1                                     WhatsApp Messenger   
2                                              Instagram   
3               Messenger – Text and Video Chat for Free   
4                                         Clash of Clans   
5                Clean Master- Space Cleaner & Antivirus   
6                                         Subway Surfers   
7                                                YouTube   
8      Security Master - Antivirus, VPN, AppLock, Boo...   
9                                           Clash Royale   
10                                      Candy Crush Saga   
11           UC Browser - Fast Download Private & Secure   
12                                              Snapchat   
13       360 Security - Free Antivirus, Booster, Cleaner   
14                                        My Talking Tom   
15                                           8 Ball Pool   
16     DU Battery Saver - Battery Charger & Battery Life   
17                           BBM - Free Calls & Messages   
18     Cache Cleaner-DU Speed Booster (booster & clea...   
19                                               Twitter   
20                                       Viber Messenger   
21                                        Shadow Fight 2   
22                                         Google Photos   
23                           LINE: Free Calls & Messages   
24                                                   Pou   
25                         Skype - free IM & video calls   
26                                            Pokémon GO   
27              Minion Rush: Despicable Me Official Game   
28              Minion Rush: Despicable Me Official Game   
29                                               Yes day   
...                                                  ...   
10021  25 Mins Ako ay may lobo Etc Pinoy Kid Song Off...   
10022             Ako ay may lobo Pinoy Kid Song Offline   
10023                           Mehmet Emin Ay İlahileri   
10024                                         AY Oakmont   
10025                                          YAKALA AY   
10026                      Ay Vamos - PJ. Balvin - Piano   
10027                                           Ay Peruk   
10028                                           BA SALES   
10029  Curso Básico y Avanzado la Fe de Jesús Audio-T...   
10030                              FE Connect Drive-Tech   
10031                                        BlueDV AMBE   
10032                                             PAY DT   
10033                                       DV Portfolio   
10034                                       AV Anthony's   
10035                     Nougat Android 7 Launcher : AW   
10036                                         AW Connect   
10037                                                 AW   
10038         FE Other Disciplines Engineering Exam Prep   
10039                      Caprock Santa Fe Credit Union   
10040                     AX Timesheets App for Dynamics   
10041                                          AX Window   
10042                           INTERACTIVE CALCULUS PRO   
10043                                    DT Technologies   
10044                                         DT Freight   
10045                                  FE Connect Fhoton   
10046  Eat Right Diet (by Dt Shreya's Family Diet Cli...   
10047                             Test Application DT 02   
10048                           Santa Fe Espresso & News   
10049                                    Santa Fe Sentry   
10050                      Basic Strategy Training BJ 21   

                          Genres               Main Genre  
0                         Social                   Social  
1                  Communication            Communication  
2                         Social                   Social  
3                  Communication            Communication  
4                       Strategy

Di seguito utilizzo una "_groupby_" al fine di riuscire a ottenere il rating corrispondente a ogni singolo genere presente nel dataset, senza avere ripetizioni sulle righe.

In [54]:
rating_table.groupby('Main Genre')[['Rating']].mean().sort_values(by = 'Rating', ascending = False).head(17)

Rating
Main Genre                  
Events              4.435556
Puzzle              4.370732
Brain Games         4.358065
Art & Design        4.350000
Books & Reference   4.343275
Personalization     4.332215
Word                4.318182
Creativity          4.306250
Music & Audio       4.300000
Parenting           4.300000
Board               4.291228
Education           4.290290
Action & Adventure  4.288542
Casino              4.286486
Beauty              4.278571
Arcade              4.277838
Role Playing        4.272072

Quindi per ottenere il genere con il rating più alto usiamo la funzione "_idxmax()_", come mostrato di seguito:

In [55]:
rating_table.groupby('Main Genre')['Rating'].mean().idxmax()

'Events'

Mentre per ottenere il rating più alto, eseguiamo il comando "_max()_":

In [56]:
rating_table.groupby('Main Genre')['Rating'].mean().max()

4.435555555555556

Una volta ottenuti i rispettivi valori, volendo mettere tutto insieme, sarà sufficiente eseguire il seguente comando:

In [57]:
print("'{}' è il genere con un rating più alto, il quale è pari a : {} (arrotondato a 4 cifre significative)"
            .format(rating_table.groupby('Main Genre')['Rating'].mean().idxmax(),
            rating_table.groupby('Main Genre')['Rating'].mean().max().round(3)))

'Events' è il genere con un rating più alto, il quale è pari a : 4.436 (arrotondato a 4 cifre significative)


## <font color=000080> Exercise 11)</font> </p>
<a name="ex11"></a>

> ### <font color=0060A0> For each app, compute the approximate income, obtain as a product of number of installs and price</font> </p>

Per lo svolgimento di questo punto procedo alla ripulitura dell'attributo "_Price_" da eventuali simboli o caratteri non numerici, creando un campo "_Finished Price_", il quale viene in seguito convertito a variabile numerica (non più stringa).

In [58]:
play_store_nan['Finished Price'] = [re.sub('\$', '', price) for price in play_store_nan['Price']]
play_store_nan['Finished Price'] = play_store_nan['Finished Price'].apply(float)
play_store_nan[['App', 'Finished Price', 'Price']].sort_values(by = 'Finished Price', ascending = False).head(10)

App  Finished Price    Price
4367           I'm Rich - Trump Edition          400.00  $400.00
5362                      I Am Rich Pro          399.99  $399.99
4197             most expensive app (H)          399.99  $399.99
9934  I'm Rich/Eu sou Rico/أنا غني/我很有錢          399.99  $399.99
5369                          I am Rich          399.99  $399.99
5373                 I AM RICH PRO PLUS          399.99  $399.99
5358                         I am Rich!          399.99  $399.99
4362                         💎 I'm rich          399.99  $399.99
5359                 I am rich(premium)          399.99  $399.99
5354                     I am Rich Plus          399.99  $399.99

In alternativa, avrei potuto definire una funzione in grado di convertire l'attributo "_Price_" in una variabile numerica come appena visto. <br>
Non si avrebbe alcun cambiamento sul risultato, ma in termini di programmazione è esteticamente più appagante. <br>
Qui di seguito la funzione:
<br>

```python
price_to_float = re.compile('\D*(?P<amount>\d+\.{0,1}\d*)')

def Convert(price):
    found = price_to_float.search(price)
    if found:
        return float(found.group('amount'))
    else:
        return None

play_store_nan['Price Number'] = play_store_nan.Price.apply(Convert)
```

Procedo quindi a calcolare il reddito approssimativo, ottenibile dall'acquisto e dalla conseguente installazione delle varie applicazioni. <br>
I risultati verranno messi in una colonna denominata "_Income_".

In [59]:
play_store_nan['Income'] = play_store_nan['Finished Price'] * play_store_nan['N° of Installs']
#play_store_nan['Income'] = [re.sub('','$', str(income)) for income in play_store_nan['Income']]
play_store_nan[['App', 'Income', 'Price', 'N° of Installs']].sort_values(by = 'Income', ascending = False).head(10)

App      Income    Price  N° of Installs
2241                      Minecraft  69900000.0    $6.99        10000000
5351                      I am rich  39999000.0  $399.99          100000
5356              I Am Rich Premium  19999500.0  $399.99           50000
4034                  Hitman Sniper   9900000.0    $0.99        10000000
7417  Grand Theft Auto: San Andreas   6990000.0    $6.99         1000000
2883            Facetune - For Free   5990000.0    $5.99         1000000
5578        Sleep as Android Unlock   5990000.0    $5.99         1000000
8804            DraStic DS Emulator   4990000.0    $4.99         1000000
4367       I'm Rich - Trump Edition   4000000.0  $400.00           10000
5354                 I am Rich Plus   3999900.0  $399.99           10000

## <font color=000080> Exercise 12)</font> </p>
<a name="ex12"></a>

> ### <font color=0060A0> For each app, compute its minimum and maximum ` Sentiment_polarity `</font> </p>

In [60]:
user_reviews[['App', 'Sentiment', 'Sentiment_Polarity']].head(10)

App Sentiment  Sentiment_Polarity
0  10 Best Foods for You  Positive                1.00
1  10 Best Foods for You  Positive                0.25
2  10 Best Foods for You       NaN                 NaN
3  10 Best Foods for You  Positive                0.40
4  10 Best Foods for You  Positive                1.00
5  10 Best Foods for You  Positive                1.00
6  10 Best Foods for You  Positive                0.60
7  10 Best Foods for You       NaN                 NaN
8  10 Best Foods for You   Neutral                0.00
9  10 Best Foods for You   Neutral                0.00

Nonostante si possa scorgere dalla cella qui sopra, controlliamo la presenza o meno di valori nulli nel campo di interesse.

In [61]:
user_reviews['Sentiment_Polarity'].isna().any()

True

Anche in questo caso la presenza di valori nulli non dovrebbe crearci problemi, dal momento che con l'approccio che andremo a utilizzare, e cioè utilizzando le funzioni "_min()_" e "_max()_", questi valori verrebbero utilizzati solo nel caso in cui non ce ne fossero altri disponibili.

Ora inizio ad affrontare l'esercizio cercando come prima cosa di ottenere la "_Sentiment Polarity_" minima per ogni applicazione.<br>
Svolgo quindi i seguenti passaggi:<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; i) Eseguo una "*groupby*" sul dataset e applico la funzione "*min*";<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ii) Converto il risultato così ottenuto in un dataframe, per eseguire più facilmente il merge che dovrò effettuare più avanti;<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; iii) Rinomino le colonne, al fine di avere un risultato il più chiaro possibile.

In [62]:
min_sntmntplrt = user_reviews.groupby('App')['Sentiment_Polarity'].min()
min_sntmntplrt = pd.DataFrame(min_sntmntplrt)
min_sntmntplrt = min_sntmntplrt.rename(columns = {'Sentiment_Polarity' : 'min Sentiment Polarity'})
min_sntmntplrt.head()

min Sentiment Polarity
App                                                     
10 Best Foods for You                            -0.8000
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室                 -0.1125
11st                                             -1.0000
1800 Contacts - Lens Store                       -0.3000
1LINE – One Line with One Touch                  -0.8250

Ripeto i tre passaggi sopraindicati al fine di estrapolare la "_Sentiment Polarity_" massima per ogni applicazione.

In [63]:
max_sntmntplrt = user_reviews.groupby('App')['Sentiment_Polarity'].max()
max_sntmntplrt = pd.DataFrame(max_sntmntplrt)
max_sntmntplrt = max_sntmntplrt.rename(columns = {'Sentiment_Polarity' : 'MAX Sentiment Polarity'})
max_sntmntplrt.head()

MAX Sentiment Polarity
App                                                     
10 Best Foods for You                           1.000000
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室                0.910000
11st                                            1.000000
1800 Contacts - Lens Store                      0.838542
1LINE – One Line with One Touch                 1.000000

Eseguo quindi il merge annunciato precedentemente.

In [64]:
sentiment_polarity_mnm = min_sntmntplrt.merge(max_sntmntplrt, left_index = True, right_index = True)
sentiment_polarity_mnm.head(20)

min Sentiment Polarity  \
App                                                                          
10 Best Foods for You                                            -0.800000   
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室                                 -0.112500   
11st                                                             -1.000000   
1800 Contacts - Lens Store                                       -0.300000   
1LINE – One Line with One Touch                                  -0.825000   
2018Emoji Keyboard 😂 Emoticons Lite -sticker&gif                 -0.800000   
21-Day Meditation Experience                                     -0.265625   
2Date Dating App, Love and matching                              -0.645833   
2GIS: directory & navigator                                      -0.375000   
2RedBeans                                                        -0.800000   
2ndLine - Second Phone Number                                    -0.781250   
30 Day Fitness Challenge - Workout at Home                       -0.500000   
365Scores - Live Scores                                           0.000000   
3D Blue Glass Water Keyboard Theme                                     NaN   
3D Color Pixel by Number - Sandbox Art Coloring                        NaN   
3D Live Neon Weed Launcher                                        0.136364   
4 in a Row                                                       -0.800000   
4K Wallpapers and Ultra HD Backgrounds                           -0.700000   
591房屋交易-租屋、中古屋、新建案、實價登錄、別墅透天、公寓套房、捷運、買房賣房行情、房價房貸查詢               -0.300000   
591房屋交易-香港                                                       -0.077083   

                                                    MAX Sentiment Polarity  
App                                                                         
10 Best Foods for You                                             1.000000  
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室                                  0.910000  
11st                                                              1.000000  
1800 Contacts - Lens Store                                        0.838542  
1LINE – One Line with One Touch                                   1.000000  
2018Emoji Keyboard 😂 Emoticons Lite -sticker&gif                  1.000000  
21-Day Meditation Experience                                      0.587500  
2Date Dating App, Love and matching                               1.000000  
2GIS: directory & navigator                                       1.000000  
2RedBeans                                                         1.000000  
2ndLine - Second Phone Number                                     1.000000  
30 Day Fitness Challenge - Workout at Home                        0.900000  
365Scores - Live Scores                                           1.000000  
3D Blue Glass Water Keyboard Theme                                     NaN  
3D Color Pixel by Number - Sandbox Art Coloring                        NaN  
3D Live Neon Weed Launcher                                        1.000000  
4 in a Row                                                        1.000000  
4K Wallpapers and Ultra HD Backgrounds                            0.750000  
591房屋交易-租屋、中古屋、新建案、實價登錄、別墅透天、公寓套房、捷運、買房賣房行情、房價房貸查詢                0.700000  
591房屋交易-香港                                                        0.700000

Se si volessero eliminare i valori nulli, mi basterà eseguire il pezzetto di codice sotto riportato. <br>
L'eliminazione di tali valori non è strettamente necessaria, ma garantisce la restituzione di un output più pulito.

In [65]:
sentiment_polarity_mnm[sentiment_polarity_mnm.notnull().any(axis = 1)].head(20)

min Sentiment Polarity  \
App                                                                          
10 Best Foods for You                                            -0.800000   
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室                                 -0.112500   
11st                                                             -1.000000   
1800 Contacts - Lens Store                                       -0.300000   
1LINE – One Line with One Touch                                  -0.825000   
2018Emoji Keyboard 😂 Emoticons Lite -sticker&gif                 -0.800000   
21-Day Meditation Experience                                     -0.265625   
2Date Dating App, Love and matching                              -0.645833   
2GIS: directory & navigator                                      -0.375000   
2RedBeans                                                        -0.800000   
2ndLine - Second Phone Number                                    -0.781250   
30 Day Fitness Challenge - Workout at Home                       -0.500000   
365Scores - Live Scores                                           0.000000   
3D Live Neon Weed Launcher                                        0.136364   
4 in a Row                                                       -0.800000   
4K Wallpapers and Ultra HD Backgrounds                           -0.700000   
591房屋交易-租屋、中古屋、新建案、實價登錄、別墅透天、公寓套房、捷運、買房賣房行情、房價房貸查詢               -0.300000   
591房屋交易-香港                                                       -0.077083   
7 Cups: Anxiety & Stress Chat                                    -0.172222   
7 Day Food Journal Challenge                                      0.070000   

                                                    MAX Sentiment Polarity  
App                                                                         
10 Best Foods for You                                             1.000000  
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室                                  0.910000  
11st                                                              1.000000  
1800 Contacts - Lens Store                                        0.838542  
1LINE – One Line with One Touch                                   1.000000  
2018Emoji Keyboard 😂 Emoticons Lite -sticker&gif                  1.000000  
21-Day Meditation Experience                                      0.587500  
2Date Dating App, Love and matching                               1.000000  
2GIS: directory & navigator                                       1.000000  
2RedBeans                                                         1.000000  
2ndLine - Second Phone Number                                     1.000000  
30 Day Fitness Challenge - Workout at Home                        0.900000  
365Scores - Live Scores                                           1.000000  
3D Live Neon Weed Launcher                                        1.000000  
4 in a Row                                                        1.000000  
4K Wallpapers and Ultra HD Backgrounds                            0.750000  
591房屋交易-租屋、中古屋、新建案、實價登錄、別墅透天、公寓套房、捷運、買房賣房行情、房價房貸查詢                0.700000  
591房屋交易-香港                                                        0.700000  
7 Cups: Anxiety & Stress Chat                                     0.812500  
7 Day Food Journal Challenge                                      0.343333

***

<a name="paragraph3"></a>

## <font color=BF4100> Conclusion </font>

<a name="subparagraph4"></a>

### <p>&#9656;  <font color=E26200> First "*time*" ending </font> </p>

Dal momento che mi sono reso conto che la presenza di visualizzazioni rallenta molto il tempo di esecuzione del codice, decido di eseguire una prima interruzione della funzione "_time_" a questo punto del Notebook.

In [66]:
finish_1 = time.time()

print('Il tempo di esecuzione del codice è stato di ' + str(finish_1-begin) + ' secondi.')

Il tempo di esecuzione del codice è stato di 7.939036846160889 secondi.


La relativa lentezza è probabilmente causata dalle immagini e dalle molte celle markwdown di commento inserite all'inizio e lungo tutto il foglio di lavoro.

<a name="subparagraph5"></a>

### <p>&#9656;  <font color=E26200> Some Data Visualizations</font> </p>

In [67]:
from plotly import __version__
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
import plotly.graph_objs as go
import cufflinks as cf
import seaborn as sns
init_notebook_mode(connected = True)
cf.go_offline()

In [68]:
cxa = play_store_nan.groupby("Category", as_index = False, sort = False)["App"].count()
cxa = cxa.sort_values(by = "App", ascending = False).head(20)

In [69]:
cxa.iplot(kind = "bar", x = "Category", bargap = 0.3, bargroupgap = 0.05, opacity = 0.60, color = '#00603C',
           layout = go.Layout(autosize = False, width = 950, height = 650,
                              title = 'Top 20 Category per n° of Apps',
                              xaxis = dict(title = 'Category'), yaxis = dict(title = 'Numerosity')))

In [70]:
cxar = play_store.groupby("Category", as_index = False)["Rating"].mean()
cxar = cxar.sort_values(by = "Rating", ascending = False).head(10)

In [71]:
cxar.iplot(kind = "bar", x = "Category", bargap = 0.3, bargroupgap = 0.05, opacity = 0.60, color = "#8000000",
           layout = go.Layout(autosize = False, width = 850, height = 650, 
                              title = 'Top 10 Category by Average Rating',
                              yaxis = dict(title = 'Average Rating')))

<a name="subparagraph6"></a>

### <p>&#9656;  <font color=E26200> Second "*time*" ending </font> </p>

Qui di seguito riportiamo il tempo di esecuzione complessivo del Notebook.

In [72]:
finish_2 = time.time()

print('Il tempo di esecuzione del codice, compreso di visualizzazioni, è stato di ' + str(finish_2-begin) + ' secondi.')

Il tempo di esecuzione del codice, compreso di visualizzazioni, è stato di 11.738164901733398 secondi.


Come già premesso nella sezione "First "_time_" ending", il rallentamento maggiore è dato dall'inserimento delle visualizzazioni.

In [73]:
print('Più precisamente, la differenza tra i due tempi di esecuzione ammonta a ' + str(finish_2 - finish_1) + ' secondi.')

Più precisamente, la differenza tra i due tempi di esecuzione ammonta a 3.7991280555725098 secondi.


***

<tr><td><br><br><br><br><br>[<img src = "img/thumbnail.png", width = 20, height = 20>](https://www.kaggle.com/alessandroborroni)</td> <td><br><br><br><br><br>[<img src = "img/Twitter_bird_logo.png", width = 18, height = 18>](https://twitter.com/malborroni)</td> <td><br><br><br><br><br>[<img src = "img/GitHub.png", width = 20, height = 20>](https://github.com/malborroni)</td></tr>